examples_tests/23.Autoexposure/main.cpp

```diff
@@ -4,20 +4,771 @@
+			layoutTransBarrier.dstQueueFamilyIndex = ~0u;
+			layoutTransBarrier.image = tonemapperOutImageViews[acquiredNextFBO]->getCreationParameters().image;
+			layoutTransBarrier.subresourceRange.aspectMask = video::IGPUImage::EAF_COLOR_BIT;
+			layoutTransBarrier.subresourceRange.baseArrayLayer = 0u;
+			layoutTransBarrier.subresourceRange.layerCount = imageCreationParams.arrayLayers;
+			layoutTransBarrier.subresourceRange.baseMipLevel = 0u;
+			layoutTransBarrier.subresourceRange.levelCount = imageCreationParams.mipLevels;
+		}
+
+		ToneMapperClass::dispatchHelper(
+			cb.get(),
+			tonemapperOutImageViews[acquiredNextFBO].get(),
+			asset::EPSF_TOP_OF_PIPE_BIT,
+			0u, nullptr,
+			asset::EPSF_FRAGMENT_SHADER_BIT,
+			1u, &layoutTransBarrier);
+
+		// This can probably be sucked into ToneMapperClass::dispatchHelper
+		if (usingTemporalAdapatation)
+		{
+			auto thisPresentStamp = std::chrono::high_resolution_clock::now();
+			auto microsecondsElapsedBetweenPresents = std::chrono::duration_cast<std::chrono::microseconds>(thisPresentStamp - lastPresentStamp);
+			lastPresentStamp = thisPresentStamp;
+
+			tonemappingParams.setAdaptationFactorFromFrameDelta(float(microsecondsElapsedBetweenPresents.count()) / 1000000.f);
+
+			// dont override shader output
+			constexpr auto offsetPastLumaHistory = offsetof(decltype(tonemappingParams), lastFrameExtraEVAsHalf) + sizeof(decltype(tonemappingParams)::lastFrameExtraEVAsHalf);
+			auto* paramPtr = reinterpret_cast<const uint8_t*>(&tonemappingParams);
+			asset::SBufferRange<video::IGPUBuffer> bufferRange = {};
+			bufferRange.buffer = toneParamsLumaOutputBuffer[acquiredNextFBO];
+			bufferRange.offset = offsetPastLumaHistory;
+			bufferRange.size = sizeof(tonemappingParams) - bufferRange.offset;
+
+			// memory dependency to ensure all the compute shader read for the tonemapping dispatch
+			// has finished before we try to transfer new stuff to the buffer
+			video::IGPUCommandBuffer::SBufferMemoryBarrier prevParamsReadBarrier = {};
+			prevParamsReadBarrier.barrier.srcAccessMask = asset::EAF_SHADER_READ_BIT;
+			prevParamsReadBarrier.barrier.dstAccessMask = asset::EAF_TRANSFER_WRITE_BIT;
+			prevParamsReadBarrier.srcQueueFamilyIndex = ~0u;
+			prevParamsReadBarrier.dstQueueFamilyIndex = ~0u;
+			prevParamsReadBarrier.buffer = bufferRange.buffer;
+			prevParamsReadBarrier.offset = bufferRange.offset;
+			prevParamsReadBarrier.size = bufferRange.size;
+
+			cb->pipelineBarrier(
+				asset::EPSF_COMPUTE_SHADER_BIT,
+				asset::EPSF_TRANSFER_BIT,
+				asset::EDF_BY_REGION_BIT,
+				0u, nullptr,
+				1u, &prevParamsReadBarrier,
+				0u, nullptr);
+
+			uint32_t waitSemCount = 0u;
+			video::IGPUSemaphore *const* waitSem = nullptr;
+			const asset::E_PIPELINE_STAGE_FLAGS* waitStages = nullptr;
+			utilities->updateBufferRangeViaStagingBuffer(
+				cb.get(),
+				nullptr,
+				graphicsComputeQueue,
+				bufferRange,
+				paramPtr + offsetPastLumaHistory,
+				waitSemCount, waitSem, waitStages);
+		}
```

Comment:
ugh, use `cb->updateBuffer` instead of this `updateBufferRangeViaStagingBuffer` for a few byte update is like bringing nuke to a gunfight